pkg/ingestor/parser/parser.go

```diff
@@ -0,0 +1,286 @@
+		foundIdentities:       []assembler.IdentityNode{},
+		foundSubjectArtifacts: []assembler.ArtifactNode{},
+		foundAttestations:     []assembler.AttestationNode{},
+		foundBuilders:         []assembler.BuilderNode{},
+		foundDependencies:     []assembler.ArtifactNode{},
+	}
+}
+
+// ParseDocumentTree takes the DocumentTree and create graph inputs (nodes and edges) per document node
+func ParseDocumentTree(docTree processor.DocumentTree) ([]assembler.AssemblerInput, error) {
+
+	assemblerinputs := []assembler.AssemblerInput{}
+	docTreeBuilder := newDocTreeBuilder()
+	err := docTreeBuilder.parse(docTree)
+	if err != nil {
+		return []assembler.AssemblerInput{}, err
+	}
+	for _, builder := range docTreeBuilder.graphBuilders {
+		assemblerinput := builder.createAssemblerInput(docTreeBuilder.identities)
+		assemblerinputs = append(assemblerinputs, assemblerinput)
+	}
+
+	return assemblerinputs, nil
+}
+
+func (b *graphBuilder) createAssemblerInput(foundIdentities []assembler.IdentityNode) assembler.AssemblerInput {
+	assemblerinput := assembler.AssemblerInput{
+		Nodes: b.createNodes(),
+		Edges: b.createEdges(foundIdentities),
+	}
+	return assemblerinput
+}
+
+func (b *graphBuilder) createEdges(foundIdentities []assembler.IdentityNode) []assembler.GuacEdge {
+	edges := []assembler.GuacEdge{}
+	for _, i := range foundIdentities {
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.IdentityForEdge{IdentityNode: i, AttestationNode: a})
+		}
+	}
+	for _, s := range b.foundSubjectArtifacts {
+		for _, b := range b.foundBuilders {
+			edges = append(edges, assembler.BuiltByEdge{ArtifactNode: s, BuilderNode: b})
+		}
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.AttestationForEdge{AttestationNode: a, ArtifactNode: s})
+		}
+		for _, d := range b.foundDependencies {
+			edges = append(edges, assembler.DependsOnEdge{ArtifactNode: s, Dependency: d})
+		}
+	}
+	return edges
+}
+
+func (b *graphBuilder) createNodes() []assembler.GuacNode {
+	nodes := []assembler.GuacNode{}
+	for _, i := range b.foundIdentities {
+		nodes = append(nodes, i)
+	}
+	for _, s := range b.foundSubjectArtifacts {
+		nodes = append(nodes, s)
+	}
+	for _, a := range b.foundAttestations {
+		nodes = append(nodes, a)
+	}
+	for _, d := range b.foundDependencies {
+		nodes = append(nodes, d)
+	}
+	for _, b := range b.foundBuilders {
+		nodes = append(nodes, b)
+	}
+	return nodes
+}
+
+func (t *docTreeBuilder) parse(root processor.DocumentTree) error {
+	builder := newGraphBuilder()
+	err := builder.parserHelper(root.Document)
+	if err != nil {
+		return err
+	}
+
+	t.graphBuilders = append(t.graphBuilders, builder)
+	t.identities = append(t.identities, builder.foundIdentities...)
+
+	if len(root.Children) == 0 {
+		return nil
+	}
+
+	for _, c := range root.Children {
+		err := t.parse(c)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (b *graphBuilder) parserHelper(doc *processor.Document) error {
```

Comment:
i think it may make sense to do the document switch statement here and then call the get functions within that ? trying to imagine what it would like as a pluggable interface. And move the SLSA specific stuff to `parser_slsa.go`

Reply:
broke the parser out. Might need to discuss on the predicate dictionary implementation.